src/components/ha-cover-controls.ts

```diff
@@ -0,0 +1,138 @@
+
+  @property() public stateObj!: HassEntity;
+
+  protected render() {
+    const entityObj = new CoverEntity(this.hass, this.stateObj);
+
+    return html`
+      <div class="state">
+        <ha-icon-button
+          class=${classMap({
+            invisible: entityObj.supportsOpen,
+          })}
+          label="Open cover"
+          .icon=${this.computeOpenIcon()}
+          @click=${this.onOpenTap}
+          .disabled=${this.computeOpenDisabled()}
+        ></ha-icon-button>
+        <ha-icon-button
+          class=${classMap({
+            invisible: entityObj.supportsStop,
+          })}
+          label="Stop the cover from moving"
+          icon="hass:stop"
+          @click=${this.onStopTap}
+          .disabled=${this.computeStopDisabled()}
+        ></ha-icon-button>
+        <ha-icon-button
+          class=${classMap({
+            invisible: entityObj.supportsClose,
+          })}
+          label="Close cover"
+          .icon=${this.computeCloseIcon()}
+          @click=${this.onCloseTap}
+          .disabled=${this.computeClosedDisabled()}
+        ></ha-icon-button>
+      </div>
+    `;
+  }
+
+  static get styles(): CSSResult {
+    return css`
+      .state {
+        white-space: nowrap;
+      }
+      [invisible] {
+        visibility: hidden !important;
+      }
+    `;
+  }
+
+  private computeOpenIcon(): string {
+    switch (this.stateObj.attributes.device_class) {
+      case "awning":
+      case "door":
+      case "gate":
+        return "hass:arrow-expand-horizontal";
+      default:
+        return "hass:arrow-up";
+    }
+  }
+
+  private computeCloseIcon(): string {
+    switch (this.stateObj.attributes.device_class) {
+      case "awning":
+      case "door":
+      case "gate":
+        return "hass:arrow-collapse-horizontal";
+      default:
+        return "hass:arrow-down";
+    }
+  }
+
+  private computeStopDisabled(): boolean {
+    if (this.stateObj.state === UNAVAILABLE) {
+      return true;
+    }
+    return false;
+  }
+
+  private computeOpenDisabled(): boolean {
+    if (this.stateObj.state === UNAVAILABLE) {
+      return true;
+    }
+    const assumedState = this.stateObj.attributes.assumed_state === true;
+    const entityObj = new CoverEntity(this.hass, this.stateObj);
+    return (entityObj.isFullyOpen || entityObj.isOpening) && !assumedState;
+  }
+
+  private computeClosedDisabled(): boolean {
+    if (this.stateObj.state === UNAVAILABLE) {
+      return true;
+    }
+    const assumedState = this.stateObj.attributes.assumed_state === true;
+    const entityObj = new CoverEntity(this.hass, this.stateObj);
+    return (entityObj.isFullyClosed || entityObj.isClosing) && !assumedState;
+  }
+
+  private onOpenTap(ev): void {
+    ev.stopPropagation();
+    new CoverEntity(this.hass, this.stateObj).openCover();
+  }
+
+  private onCloseTap(ev): void {
+    ev.stopPropagation();
+    new CoverEntity(this.hass, this.stateObj).closeCover();
```

Comment:
Can we not do this in every function, but save it on the class and update it when the state changes?